tensorboard/db.py

```diff
@@ -0,0 +1,179 @@
+# Copyright 2017 The TensorFlow Authors. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""TensorBoard core database schema.
+
+Currently the only database supported is SQLite.
+
+WARNING: This module is EXPERIMENTAL. It will not be considered stable
+until data migration tools are put into place. Until that time, any
+database created with this schema will need to be deleted as new updates
+occur.
+
+## Overview
+
+TensorBoard stores data using SQL via the PEP 249 API. We accomplish
+portability by restricting ourselves to an understanding of SQL that
+works efficiently in both the tiniest database in the world, SQLite, as
+well as the largest database in the world, Spanner. This, in theory,
+should mean all other ACID SQL DBs in-between are supported as well.
+
+## Primary Keys
+
+Every table has a primary key. This is considered a SQL best practice
+and is mandated by Spanner[1].
+
+Primary keys are randomly generated because it scales better than auto
+incremented IDs. Monotonic indexes in distributed systems like Spanner
+tend to cause all operations to go to a single node[2].
+
+Primary keys are 63-bit integers. They act not only as indexes, but also
+control where the row data is stored. We are constrained to this data
+type because SQLite will only make a field the "rowid"[3] if it's
+declared "INTEGER PRIMARY KEY" and SQLite doesn't support unsigned
+integers. Please note that INTEGER is 32-bits in MySQL and PostgreSQL.
+
+To control the locality of rows, a multi-field key will be encoded into
+the 63-bits. When this happens, the higher and lower bits will be random
+and the middle bits will encode the second field, e.g. timestamp for a
```

Comment:
That will be documented in each individual table. This overview is a tiny bit more abstract. I imagine some might be 16 + 32 + 15 where 16 ++ 15 are the random bits and 32 is a unix timestamp.